### PR TITLE
Skip no-op delta rebuilds for root-local hash indexes

### PR DESCRIPTION
**Pull request by tnabulsi-netflix**
## Summary
- skip `HollowHashIndex` delta reindexing when the index is root-select and all match fields are direct non-reference fields on the root type
- keep the existing rebuild-on-every-delta behavior for traversing or reference-based hash indexes
- add regression tests covering skipped rebuilds, rebuilds on root changes, and unchanged behavior for non-root-local indexes

## Testing
- `./gradlew :hollow:test --tests com.netflix.hollow.core.index.HollowHashIndexTest --rerun-tasks -Pcom.netflix.gradle-predictive-test-selection.enabled=false`

### CODE REVIEW

File: hollow/src/main/java/com/netflix/hollow/core/index/HollowHashIndex.java [hunk 1]
```diff
@@ -117,6 +122,39 @@ private void reindexHashIndex() {
         this.hashStateVolatile = new HollowHashIndexState(builder);
     }
 
+    private boolean canSkipNoOpDeltaUpdate() {
```
Comment (azeng-netflix) [inline hollow/src/main/java/com/netflix/hollow/core/index/HollowHashIndex.java:125]: This part is not needed, as ordinal change is a direct signal on whether or not a hash needs to rebuilt.
One reason is that, for top level type A, if its member type B has a field change, the change would cause the type A record to have an ordinal change as well. 
This statement holds true, also because:
1. for a type, an ordinal will not be reassigned within a single cycle.
2. Hollow Data set version update via delta would apply every delta on the version chain, and would not skip a version.
3. a sub type's ordinal change would propagate all the way up to the top level type, all the record along the hierarchy would get assigned a new ordinal.

But also wonder if there are other edge cases..

Comment (azeng-netflix) [reply hollow/src/main/java/com/netflix/hollow/core/index/HollowHashIndex.java:125]: @azeng-netflix need to check if double snapshot scenario is also handled and add unit test that would check this invariant in the future.
